### PR TITLE
51-feature/table: Data Table Features and Codebase Refactor

### LINKED ISSUE #51: feature/table

### PR DESCRIPTION
**Pull request by Jervx**
This PR introduces new **Data Table features** and includes refactoring and cleanup tasks to improve the codebase.

---

### **Data Table Features**
- **Column Pinning**
  - Pin columns to the left, right, or center (default).
- **Column Resizing**
  - Enable resizing for pinned columns.
- **Column Reordering**
  - Drag-and-drop functionality to reorder columns.
- **Server-Side Data Management**
  - Filtering, sorting, and pagination handled server-side.
- **Data Export**
  - Export options for:
    - Selected rows.
    - Current page.
    - All filtered rows.
    - All unfiltered rows.
- **Delete Selected Rows**
  - Bulk deletion functionality for selected rows.

---

## **Refactor and Cleanup**
- **Icons Refactor**
  - Renamed icons for better consistency.
  - Updated imports and exports to align with the new naming convention.
- **Removed Unnecessary Files**
  - Removed `*.tsbuildinfo` files from Git tracking.
- **Codebase Formatting**
  - Reformatted the codebase for consistent styling.

---

Screenshot
image

Merging this PR will close #51

### CODE REVIEW

**Review by zalven-official — COMMENTED** [review]
To make the `syncRowHeights` and `syncFooterRowHeights` functions reusable, we can abstract out the common logic into a generalized function that handles both rows and footer rows. This way, you can call the function for different tables (rows or footer rows) while keeping the code DRY (Don't Repeat Yourself).

Here's how you can refactor it:

### Step 1: Create a generic function to synchronize heights

```tsx
const syncHeights = useCallback(
    (tableRefs: React.RefObject<HTMLElement>[], rowDataAttr: string, idAttr: string) => {
        const rowIds = new Set<string>();

        // Collect all row IDs from each table
        tableRefs.forEach((tableRef) => {
            const rows = tableRef.current?.querySelectorAll(`[${rowDataAttr}]`);
            if (rows) {
                rows.forEach((row) =>
                    rowIds.add((row as HTMLElement).dataset[idAttr]!)
                );
            }
        });

        // Sync the heights for each rowId
        rowIds.forEach((rowId) => {
            const rows = tableRefs
                .map((tableRef) =>
                    tableRef.current?.querySelector(
                        `[${rowDataAttr}="${rowId}"]`
                    )
                )
                .filter(Boolean) as HTMLElement[];

            let maxHeight = 0;
            rows.forEach((row) => {
                row.style.height = 'auto';
                maxHeight = Math.max(maxHeight, row.offsetHeight);
            });

            rows.forEach((row) => {
                row.style.height = `${maxHeight}px`;
            });
        });
    },
    []
);
```

### Step 2: Use the generic function for row heights and footer row heights

Now, you can call `syncHeights` for both row and footer row heights by passing in the appropriate table references and data attributes.

For `syncRowHeights`:

```tsx
const syncRowHeights = useCallback(() => {
    syncHeights(
        [leftTableRef, centerTableRef, rightTableRef],
        'data-row-id',
        'rowId'
    );
}, [syncHeights]);
```

For `syncFooterRowHeights`:

```tsx
const syncFooterRowHeights = useCallback(() => {
    syncHeights(
        [leftTableRef, centerTableRef, rightTableRef],
        'data-footer-row-id',
        'footerRowId'
    );
}, [syncHeights]);
```

### Explanation

1. **syncHeights** is a reusable function that accepts:
   - `tableRefs`: An array of `React.RefObject<HTMLElement>` references to your left, center, and right table containers.
   - `rowDataAttr`: The attribute that holds the row ID (`data-row-id` for rows or `data-footer-row-id` for footers).
   - `idAttr`: The data attribute key (`rowId` or `footerRowId`).

2. The function works by:
   - Collecting all the unique row IDs across all tables using the `rowDataAttr` and `idAttr` attributes.
   - Then, it calculates the maximum height for each row and sets the height for all matching rows in each table to this value.

### Advantages of this approach:
- **Reusability**: You can use this single function to handle both rows and footer rows, avoiding repetition.
- **Scalability**: You can easily extend this to support more tables or attributes in the future.
- **Maintainability**: Changes to the height synchronization logic are centralized, making the code easier to maintain.

Let me know if you need further modifications!

File: client/src/components/data-table/data-table-footer-aggregations/data-table-footer-summation.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { useMemo } from 'react'
+import { Column, Header, Table } from '@tanstack/react-table'
+
+import { cn, formatNumber } from '@/lib'
+import { IBaseCompNoChild } from '@/types'
+
+interface DataTableFooterSummationProps<TData, TValue>
+    extends IBaseCompNoChild {
+    table: Table<TData>
+    sumAll?: boolean
```
Comment (zalven-official) [inline client/src/components/data-table/data-table-footer-aggregations/data-table-footer-summation.tsx:10]: delete all

File: client/src/components/data-table/data-table-footer-aggregations/data-table-footer-summation.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    sumAll = false,
+    totalLabel = 'Total',
+}: DataTableFooterSummationProps<TData, TValue>) => {
+    const sum = useMemo(() => {
+        const { pageIndex, pageSize } = table.getState().pagination
+
+        const startRow = pageIndex * pageSize
+        const endRow = startRow + pageSize
+
+        const rows = sumAll
```
Comment (zalven-official) [inline client/src/components/data-table/data-table-footer-aggregations/data-table-footer-summation.tsx:29]: Delete sum all

File: client/src/components/data-table/data-table-row-actions/index.tsx
Comment (zalven-official) [inline client/src/components/data-table/data-table-row-actions/index.tsx:1]: Make all function arrow something ()=>

File: client/src/components/data-table/data-table-selection-indicator.tsx
Comment (zalven-official) [inline client/src/components/data-table/data-table-selection-indicator.tsx:1]: Removed unused components

File: client/src/components/data-table/data-table-virtualize.tsx
Comment (zalven-official) [inline client/src/components/data-table/data-table-virtualize.tsx:1]: Please add comment for this unused desyo

File: client/src/components/data-table/hooks/use-datatable-state.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+export type TDataTableDisplayType = 'Default' | 'Full'
+
+const useDataTableState = () => {
+    const [rowSelection, setRowSelection] = useState({})
+    const [sorting, setSorting] = useState<SortingState>([])
+    const [columnVisibility, setColumnVisibility] = useState({})
+    const [pagination, setPagination] = useState<PaginationState>({
+        pageIndex: 0,
+        pageSize: 10,
```
Comment (zalven-official) [inline client/src/components/data-table/hooks/use-datatable-state.ts:12]: make props and default page size in constant/table.ts

File: client/src/components/data-table/data-table-column-header/column-filters/multi-select-filter.tsx
Comment (zalven-official) [inline client/src/components/data-table/data-table-column-header/column-filters/multi-select-filter.tsx:1]: Make this global

**Review by zalven-official — CHANGES_REQUESTED** [review]
To make the `syncRowHeights` and `syncFooterRowHeights` functions reusable, we can abstract out the common logic into a generalized function that handles both rows and footer rows. This way, you can call the function for different tables (rows or footer rows) while keeping the code DRY (Don't Repeat Yourself).

Here's how you can refactor it:

### Step 1: Create a generic function to synchronize heights

```tsx
const syncHeights = useCallback(
    (tableRefs: React.RefObject<HTMLElement>[], rowDataAttr: string, idAttr: string) => {
        const rowIds = new Set<string>();

        // Collect all row IDs from each table
        tableRefs.forEach((tableRef) => {
            const rows = tableRef.current?.querySelectorAll(`[${rowDataAttr}]`);
            if (rows) {
                rows.forEach((row) =>
                    rowIds.add((row as HTMLElement).dataset[idAttr]!)
                );
            }
        });

        // Sync the heights for each rowId
        rowIds.forEach((rowId) => {
            const rows = tableRefs
                .map((tableRef) =>
                    tableRef.current?.querySelector(
                        `[${rowDataAttr}="${rowId}"]`
                    )
                )
                .filter(Boolean) as HTMLElement[];

            let maxHeight = 0;
            rows.forEach((row) => {
                row.style.height = 'auto';
                maxHeight = Math.max(maxHeight, row.offsetHeight);
            });

            rows.forEach((row) => {
                row.style.height = `${maxHeight}px`;
            });
        });
    },
    []
);
```

### Step 2: Use the generic function for row heights and footer row heights

Now, you can call `syncHeights` for both row and footer row heights by passing in the appropriate table references and data attributes.

For `syncRowHeights`:

```tsx
const syncRowHeights = useCallback(() => {
    syncHeights(
        [leftTableRef, centerTableRef, rightTableRef],
        'data-row-id',
        'rowId'
    );
}, [syncHeights]);
```

For `syncFooterRowHeights`:

```tsx
const syncFooterRowHeights = useCallback(() => {
    syncHeights(
        [leftTableRef, centerTableRef, rightTableRef],
        'data-footer-row-id',
        'footerRowId'
    );
}, [syncHeights]);
```

### Explanation

1. **syncHeights** is a reusable function that accepts:
   - `tableRefs`: An array of `React.RefObject<HTMLElement>` references to your left, center, and right table containers.
   - `rowDataAttr`: The attribute that holds the row ID (`data-row-id` for rows or `data-footer-row-id` for footers).
   - `idAttr`: The data attribute key (`rowId` or `footerRowId`).

2. The function works by:
   - Collecting all the unique row IDs across all tables using the `rowDataAttr` and `idAttr` attributes.
   - Then, it calculates the maximum height for each row and sets the height for all matching rows in each table to this value.

### Advantages of this approach:
- **Reusability**: You can use this single function to handle both rows and footer rows, avoiding repetition.
- **Scalability**: You can easily extend this to support more tables or attributes in the future.
- **Maintainability**: Changes to the height synchronization logic are centralized, making the code easier to maintain.

Let me know if you need further modifications!

**Comment by zalven-official** [comment]
Add global search and search injectables 

image
image
image

**Comment by zalven-official** [comment]
image
Applying filter

**Review by zalven-official — APPROVED** [review]
LGTM!

**Comment by Jervx** [comment]
Augh shems

**Comment by Jervx** [comment]
close #51

**Pull request merged**